File: ios/CashuWallet/Core/MintQuoteDomain.swift
import Foundation

/// Pure mint-quote domain rules (Android `MintQuoteDomain.kt` parity) —
/// extracted for unit testing.
enum MintQuoteDomain {
    /// Payer decoders render control characters, including newlines, as replacement glyphs.
    static func normalizedOfferDescription(_ raw: String?) -> String? {
        let printable = (raw ?? "").unicodeScalars.compactMap { scalar -> String? in
            if CharacterSet.whitespacesAndNewlines.contains(scalar) { return " " }
            return scalar.properties.generalCategory == .control ? nil : String(scalar)
        }.joined()
        let normalized = String(printable.split(separator: " ").joined(separator: " ").prefix(640))
        return normalized.isEmpty ? nil : normalized
    }

    /// True when any NUT-04 bolt12 method advertises `description: true`.
    /// Null or false on every bolt12 method (or no bolt12 method) fails closed.
    static func reportsBolt12MintDescription(
        methods: [(method: PaymentMethodKind?, description: Bool?)]
    ) -> Bool {
        methods.contains { $0.method == .bolt12 && $0.description == true }
    }

    /// Exact-match rule for reusable-offer reuse: an amountless BOLT12 quote
    /// at the active mint and requested unit whose locally stored memo equals
    /// the requested description (nil → only the plain, description-less
    /// offer). CDK never returns offer descriptions, so the memo stored locally
    /// by quote id is the only record — and exact matching keeps reuse
    /// unambiguous once several amountless offers exist (offers are immutable,
    /// so a changed description always mints a fresh one).
    static func isReusableAmountlessOffer(
        paymentMethod: PaymentMethodKind?,
        isAmountless: Bool,
        quoteMintUrl: String,
        quoteUnit: String,
        activeMintUrl: String,
        unit: String,
        storedMemo: String?,
        description: String?
    ) -> Bool {
        paymentMethod == .bolt12 &&
            isAmountless &&
            quoteMintUrl == activeMintUrl &&
            quoteUnit.lowercased() == unit.lowercased() &&
            storedMemo == description
    }
}

File: android/app/src/main/java/com/cashu/me/Core/MintQuoteDomain.kt
package com.cashu.me.Core

import com.cashu.me.Models.MintQuoteInfo
import com.cashu.me.Models.MintQuoteState
import com.cashu.me.Models.PaymentMethodKind

internal const val LOCAL_NEVER_EXPIRES_EPOCH_SECONDS: Long = 253_402_300_799L

internal fun mintQuoteLocalStorageExpiry(
    expiryEpochSeconds: Long,
    paymentMethod: PaymentMethodKind,
): Long =
    if (paymentMethod == PaymentMethodKind.Bolt12 && expiryEpochSeconds == 0L) {
        LOCAL_NEVER_EXPIRES_EPOCH_SECONDS
    } else {
        expiryEpochSeconds
    }

internal fun mintQuoteDisplayExpiry(expiryEpochSeconds: Long?): Long? =
    expiryEpochSeconds?.takeIf { it > 0 && it != LOCAL_NEVER_EXPIRES_EPOCH_SECONDS }

internal fun mintQuoteAmountForDomain(
    quoteAmount: Long?,
    fallbackAmount: Long?,
    amountPaid: Long,
    amountIssued: Long,
): Long? =
    quoteAmount
        ?: amountPaid.takeIf { it > 0 }
        ?: fallbackAmount?.takeIf { it > 0 }
        ?: amountIssued.takeIf { it > 0 }

internal fun mintQuoteStateForDomain(
    paymentMethod: PaymentMethodKind,
    storedState: MintQuoteState,
    amountPaid: Long,
    amountIssued: Long,
): MintQuoteState {
    if (amountPaid > 0 && amountIssued >= amountPaid) return MintQuoteState.Issued
    if (amountPaid > amountIssued) return MintQuoteState.Paid
    if (paymentMethod != PaymentMethodKind.Bolt11) return MintQuoteState.Pending
    return storedState
}

/**
 * Finds the long-lived, amountless BOLT12 offer for one mint wallet, matching
 * the requested [description] exactly (null → only the plain, description-less
 * offer). BOLT12 quotes remain in CDK's unissued list even after payments, so
 * deliberately do not filter by quote state here. The description match keeps
 * reuse unambiguous once several amountless offers exist (offers are
 * immutable, so a changed description always mints a fresh one).
 */
internal fun findExistingAmountlessBolt12Offer(
    quotes: List<MintQuoteInfo>,
    mintUrl: String,
    unit: String,
    description: String? = null,
): MintQuoteInfo? = quotes.firstOrNull { quote ->
    quote.paymentMethod == PaymentMethodKind.Bolt12 &&
        quote.isAmountless &&
        quote.mintUrl == mintUrl &&
        quote.unit.equals(unit, ignoreCase = true) &&
        quote.description == description
}
